Drop pyopenssl dependency

sergei-maertens opened the pull request:
~~Since cryptography 42.0.0 there is support for x509 chain validation~~

Chain validation is nuked, the problem that needed solving was not clear enough so the solution was a solution for a problem that didn't exist :)

sergei-maertens committed:
:recycle: Refactor private/public key format validation

Instead of using PyOpenSSL, use the cryptography API directly.

sergei-maertens committed:
:recycle: Refactor (public) certificate processing to use cryptography

Rather than PyOpenSSL. Turns out this leads to are more
pleasant Python API too :tada:.

sergei-maertens committed:
:recycle: Refactor certificate/key pair validation to use cryptography

... rather than PyOpenSSL.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -52,64 +59,52 @@ def __str__(self):
     @property
     def serial_number(self) -> str:
-        x509sn: int = self._certificate.get_serial_number()
+        x509sn = self._certificate.serial_number
```
Type is correctly inferred from cryptography `Certificate` class :)

Viicos replied:
The line below can be simplified as well :)

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -52,64 +59,52 @@ def __str__(self):
     def serial_number(self) -> str:
-        x509sn: int = self._certificate.get_serial_number()
+        x509sn = self._certificate.serial_number
         sn: str = hex(x509sn)[2:].upper()
         bytes = (sn[i : i + 2] for i in range(0, len(sn), 2))
         return ":".join(bytes)
```
the `str` representation of the serial number produces this but without colons, so I've left the implementation as is. Relying on the (undocumented) `str`/`repr` doesn't seem like a great idea.

Viicos replied:
> the `str` representation of the serial number produces this but without colons

I assume you are talking about the `pyOpenSSL` repr, as `.serial_number` is currently an int with cryptography.

Would have been great if `cryptography` offered a repr but this ":".join format doesn't seem to be standardized anyway.

sergei-maertens replied:
no, this was the cryptography variant in an interactive debugger. it threw me off when I thought it already returned string, but I was wrong.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -52,64 +59,52 @@ def __str__(self):
-        context = SSL.Context(SSL.TLSv1_2_METHOD)
-        context.use_privatekey(self._private_key)
-        context.use_certificate(self._certificate)
+        key_pubkey = self._private_key.public_key()
+        cert_pubkey = self._certificate.public_key()
+        return key_pubkey == cert_pubkey
```
this seems almost to good/simple to be true...

sergei-maertens commented on simple_certmanager/validators.py:
```diff
@@ -11,30 +14,35 @@ class PKIValidatorBase:
         try:
             self.validate(value.read())
-        except crypto.Error:
+        except ValueError:
```
cryptography raises `ValueError` if it doesn't look like a valid PEM thing.

sergei-maertens commented on tests/test_certificates.py:
```diff
@@ -29,7 +29,10 @@ def test_calculated_properties(self):
             )
 
-        self.assertEqual(datetime(2023, 2, 21, 14, 26, 51), certificate.expiry_date)
+        self.assertEqual(
+            certificate.expiry_date,
+            datetime(2023, 2, 21, 14, 26, 51, tzinfo=timezone.utc),
+        )
```
Slight change in behaviour because now we have aware datetimes instead of naive. The original was wrong depending on the timezone/DST etc., since the UTC datetime was being interpreted as a naive datetime, which is off either 1 or 2 hours in NL.

sergei-maertens committed:
:recycle: Refactor the certificate chain validation

Since cryptography 42.0.0 the functionality for this is present, so
we no longer need PyOpenSSL.

sergei-maertens commented on simple_certmanager/utils.py:
```diff
@@ -1,67 +1,82 @@
+    # exact host name for the chain validation since we don't know which hosts will be
+    # connected to with this certificate - that only happens at runtime. So, we use the
+    # first entry.
+    # XXX: this can crash if the SAN extension is not present or no DNS names are
+    # specified - it's unclear if those are situations we need to support or not.
+    # Probably the validation of the (leaf) certificate itself should enforce the
+    # presence of this information.
```
Two admin tests are now failing on this, I'm not sure how to handle this yet :thinking:

sergei-maertens commented:
@CharString I'll follow up on the test failures next week, but can you take a look at the refactor? Especially the certificate building in the fixtures required some fiddling around because cryptography has some requirements to be able to validate chains of certificates. Notably, the following extensions are required:

* Basic Constraints
* Key Usage
* Subject Key Identifier
* Authority Key Identifier

I cross-checked this with a Let's Encrypt issued certificate + intermediate and what we generate to get the happy flow test_utils.py tests to pass.

The failing tests are because of test certificates generated with openssl that don't match these requirements (SAN is missing, for example)

sergei-maertens committed:
:pencil: Document more test cert/key usage places

sergei-maertens committed:
:sparkles: Restrict chain validation to server certificates only

Since these are used for server verification, we check the presence
of SAN extension/extract the DNS name to validate the chain against.

sergei-maertens committed:
:fire: Remove certificate chain validation

This feature was badly specced and too vague to fix the problems
encountered in mTLS, and its implementation only causes more
confusion.

To be clear, the certificate model here serves to verify the
identity of a server (by uploading a cert-only certificate, without
private key) or to provide client-authentication through a private
key + public cert pair.

When using server certificate, this is typically passed to requests
through the 'verify' kwarg, with the file to the certificate. That
makes the specified certificate the trust store (so you upload
the certificate of the host directly, or a root or any intermediate
in the server certificate chain). It is a way to deal with root CAs
that are not in the certifi bundle.

The other, more common (?) usage is to configure the client certificate
in an mTLS set-up, where 'we' are the client and present our
certificate (and intermediates) to the server, which validates this
against its own root trust store. We can not validate this chain,
as we don't know which root certificates are trusted by the server.
The only way to actually test this, is by making a TLS connection
with our key/certificate and checking if that is accepted by the
server. This functionality will land in the future.

sergei-maertens commented on setup.cfg:
```diff
@@ -37,8 +37,7 @@ python_requires = >=3.7
     django-privates >= 1.5
-    pyopenssl >= 22
-    cryptography >= 35
+    cryptography >= 35.0.0
```
we no longer *require* the latest version, but it's still wise to keep dependencies updated. I've reverted the version bump.
